Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -94,4 +102,24 @@ void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& ori
     }
 }
 
+void WebPermissionController::permissionMayHaveChanged(WebCore::PermissionName permissionName, const WebCore::SecurityOriginData& topOrigin)
+{
+    for (auto& observer : m_observers) {
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
+            auto& context = observer.context();
+            auto source = WebCore::Permissions::sourceFromContext(context);
+            // FIXME: Add support for service workers and shared workers.
+            ASSERT(source && (*source == WebCore::PermissionQuerySource::Window || *source == WebCore::PermissionQuerySource::DedicatedWorker));
+            auto* page = downcast<WebCore::Document>(context).page();
```

Comment:
You're right, the downcast would need to be to DedicatedWorkerGlobal scope. I think we do support dedicated workers though right? Because they have a page?

Reply:
No, you have a main thread check in PermissionStatus when registering/unregistering the observer. For dedicated workers, this main thread check would fail and thus the worker's PermissionStatus currently don't register as observers.

Reply:
Ah yes, that's correct. I'll make the change to indicate that this patch is only adding support for Window contexts.